### PR TITLE
Issue114: State Tutorial

### PR DESCRIPTION
**Pull request by agubler**
Updates the code for the state tutorial, takes the initial words from @vansimke and addresses as many comments from @dylans as I could.

Supersedes https://github.com/dojo/dojo.io/pull/145

### CODE REVIEW

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+---
+
+{% section 'first' %}
+
+# State management
+
+## Overview
+
+Modern web application today are required to manage complex state models which can involve fetching data from a remote service or multiple widget requiring the same slices of state. While Dojo 2's widgets can manage application state, encapsulation and a clean separation of concerns may be lost if widgets manage their own visual representations, listen for interactions from the user, manage their children, and keep track of state information. Additionally, using widgets to pass state through an application often forces the widgets to be aware of state information for the sole purpose of passing that data down to their children. To allow widgets to remain focused on their primary roles of providing a visual representation of themselves and listening for user interactions, Dojo 2 provides `Containers` and `Injectors` that are designed to coordinate an application's external state and connect and map this state to properties only for the widgets that it is required.
```
Comment (agubler) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: Perhaps we need to be clearer that side-loading properties from external state isn't always the best way to immediately go. 

Up lifting state to the highest common widgets is a perfectly reasonable pattern for less complicated applications, do we have a tutorial that covers that explicitly?

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 2]
```diff
@@ -24,29 +24,27 @@ This tutorial assumes that you have gone through the [beginner tutorial](../001_
...
+The code begins by importing some modules, including the `WorkerProperties` and `WorkerFormData` interfaces defined in the `Worker` and `WorkerForm` modules. These two interfaces define the state that the `ApplicationContext` will track.
 
-After the imports, we have defined the `defaultWorkerForm`. This is identical to the `defautlForm` from the `App` module and is used for the same purpose - to reset the form data when a new `Worker` is created.
+The `ApplicationContext` class contains the state information. `ApplicationContext` has two private fields, `_workerData` and `_formData`, which contain the state, and two accessor methods to retrieve these fields Finally, notice the two public methods that `ApplicationContext` exposes:
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: these fields Finally, -> these fields. Finally,

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 3]
```diff
@@ -24,29 +24,27 @@ This tutorial assumes that you have gone through the [beginner tutorial](../001_
...
 
-The `ApplicationContext` class contains the state information. It contains two private fields, `_workerData` and `_formData` that will contain the state as well as two accessor methods to retrieve those fields. Finally, notice the two public methods that `ApplicationContext` exposes:
+{% include_codefile 'demo/finished/biz-e-corp/src/ApplicationContext.ts' lines:26-35 %}
 
-{% include_codefile 'demo/finished/biz-e-corp/src/ApplicationContext.ts' lines:32-41 %}
+The `formInput` method provides the same functionality as the `_onFormInput` method in the `App` class and the `submitForm` method is analogous `_addWorker` from the `App` class. The implementations vary slightly as the `ApplicationContext` has dedicated fields to store the state information. Also, since the `ApplicationContext` is not a widget, it needs to emit an `invalidate` event to inform the application to schedule a render cycle.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: analogous `_addWorker` -> analogous to `_addWorker`

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 4]
```diff
@@ -58,31 +56,35 @@ Currently, the application's `main` module is only responsible for creating the
 
 {% include_codefile 'demo/initial/biz-e-corp/src/main.ts' %}
 
-Now, we need to update `main` to construct and populate the `ApplicationContext` as well as register that context so that it is accessible to the rest of the application.
+Now, we need to update `main` to construct and populate the `ApplicationContext` create a `registry` for the application and use this to define the `ApplicationContext`. Finally in order to make the `registry` available within widget tree we need to pass the `registry` as a property to the `projector`.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: Now, we need to update `main` to construct and populate the `ApplicationContext` create a `registry` for the application and use this to define the `ApplicationContext`.   ... Missing a word and not particularly clear

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "Finally in order to make the `registry` available within widget tree we need to pass the `registry` as a property to the `projector`." -> "Finally, in order to make the `registry` available within widget tree, we need to pass the `registry` as a property to the `projector`."

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 5]
```diff
@@ -58,31 +56,35 @@ Currently, the application's `main` module is only responsible for creating the
...
+In a real-world application, this data would probably be loaded via a call to a web service or a local data store. To learn more, take a look at the [data stores tutorial](../comingsoon.html).
 {% endaside %}
 
-This is the same data that we used in the `App` module to initialize the `WorkerProperties`, but it is now in a more appropriate location. In general, the `main` module of an application should be concerned with initializing application-wide state. Also, as we mentioned earlier, the `App` class only needed the `WorkerProperties` because its children needed them.
+The state stored in the `ApplicationContext` is the same data used in the `App` module to initialize the `WorkerProperties`, but it is now decoupled into an isolated module that helps to understand and maintain the application. In general, the `main` module of an application should be concerned with initializing application wide state. Also, as we mentioned earlier, the `App` class only needed to manage the `WorkerProperties` state so that it could coordinate this through from and to it's children.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: it's children -> its children (not it is children)

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 6]
```diff
@@ -154,17 +152,17 @@ Also, the property and methods within `App` that are setting and managing state
...
+With this last change, the `App` class is now only nine lines of code. All of the state management logic is still part of the application, but it has been refactored out of the `App` class to create a more efficient application architecture.
 
-Notice that the `WorkerForm` and `WorkerContainer` widgets were not changed at all. This is an important thing to keep in mind when designing widgets - a widget should never be concerned about where its properties come from, that is what containers are for. By keeping the containers and widgets separate, we have helped to ensure that they each have a narrowly defined set of responsibilities, making them easier to reuse, test, and maintain.
+Notice that the `WorkerForm` and `WorkerContainer` widgets were not changed at all. This is an important thing to keep in mind when designing widgets - a widget should never be tightly coupled to the source of its properties. By keeping the containers and widgets separate, we have helped to ensure that they each have a narrowly defined set of responsibilities, creating a cleaner separation of concerns, and making the widgets and containers.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "and making the widgets and containers" making them what?

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+---
+
+{% section 'first' %}
+
+# State management
+
+## Overview
+
+Modern web application today are required to manage complex state models which can involve fetching data from a remote service or multiple widget requiring the same slices of state. While Dojo 2's widgets can manage application state, encapsulation and a clean separation of concerns may be lost if widgets manage their own visual representations, listen for interactions from the user, manage their children, and keep track of state information. Additionally, using widgets to pass state through an application often forces the widgets to be aware of state information for the sole purpose of passing that data down to their children. To allow widgets to remain focused on their primary roles of providing a visual representation of themselves and listening for user interactions, Dojo 2 provides two classes, `Container`, and `Injector` that are designed to coordinate an application's external state and connect and map this state to properties only for the widgets that it is required.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "or multiple widget requiring the same" -> "or multiple widgets requiring the same"

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: " Dojo 2 provides two classes, `Container`, and `Injector` that" ->  "Dojo 2 provides two classes, `Container` and `Injector`, that"

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: "widgets that it is required" -> something is odd with this tense, maybe reword it to be more explicit?

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+## Creating an application context
+
+{% task 'Create a class to manage application state.' %}
+
+To begin our discussion, let's review the initial version of `App`:
+
+{% include_codefile 'demo/initial/biz-e-corp/src/widgets/App.ts' %}
+
+Most of this widget is dedicated to holding and managing the `WorkerData` in the application. Notice, however, that it never actually uses that data itself. `App` is only containing the state and passing it to the children as required via properties. Up lifting state to the highest common widget in the tree is a valid pattern but as an application's state grows in size and complexity it is often desirable to decouple this from widgets. In larger applications, the `App` class would become complicated and more difficult to maintain due to the additional state information that it would be required track. Since the state information is not a primary concern of the `App` class, let's refactor it out of `App` and into a new `ApplicationContext` class.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "Up lifting state" -> should be "Uplifting state", though perhaps elevating or raising is a better choice than uplifting?

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: "pattern but as an application's state grows in size and complexity it is often" -> "pattern, but as an application's state grows in size and complexity, it is often"

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: "required track" -> "required to track"

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+The code begins by importing some modules, including the `WorkerProperties` and `WorkerFormData` interfaces defined in the `Worker` and `WorkerForm` modules. These two interfaces define the shape of state that the `ApplicationContext` manages.
+
+The `ApplicationContext` class contains the state information. `ApplicationContext` has two private fields, `_workerData` and `_formData`, which contain the state, and two accessor methods to retrieve these fields. Finally, notice the two public methods that `ApplicationContext` exposes:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/ApplicationContext.ts' lines:26-35 %}
+
+The `formInput` method provides the same functionality as the `_onFormInput` method in the `App` class and the `submitForm` method is analogous to `_addWorker` from the `App` class. The implementations vary slightly as the `ApplicationContext` has dedicated fields to store the state information. Also, since the `ApplicationContext` is not a widget, it cannot call `invalidate();` to schedule a re-render, the instance needs to emit an `invalidate` event that instructs associated widgets to `invalidate` themselves.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "a re-render, the instance needs to emit" -> "a re-render. Instead the instance needs to emit"

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+Notice that the `ApplicationContext` does not contain any code to load state information. Currently its only role is only to manage the application's state provided on initialization via it's `constructor`. However as the requirements for the application become more advanced the `ApplicationContext` could make requests to fetch and modify data from a remote service or local storage mechanism.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "via it's `constructor`" -> "via its `constructor`"

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+## Injectors
+
+{% task 'Register an `Injector` that will allow state to be injected into widgets.' %}
+
+Currently, the application's `main` module is only responsible for creating the `Projector`, which provides the bridge between the application code and the DOM.
+
+{% include_codefile 'demo/initial/biz-e-corp/src/main.ts' %}
+
+Now, we need to update `main` to construct and populate the an instance of `ApplicationContext`, create a `registry` and then define an `Injector` with the initialized `ApplicationContext`. Lastly, in order to make the `registry` available within widget tree we need to pass the `registry` as a property to the `projector`.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "populate the an instance" -> "populate an instance"

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "available within widget tree we need to pass" -> "available within the widget tree, we need to pass"

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+{% instruction 'Import the `ApplicationContext` module and add this code to `main`:' %}
+
+{% include_codefile 'demo/finished/biz-e-corp/src/main.ts' lines:10-29 %}
+
+{% aside 'Loading data' %}
+In a real-world application, this data would probably be loaded via a call to a web service or a local data store. To learn more, take a look at the [stores tutorial](../comingsoon.html).
+{% endaside %}
+
+The state stored in the `ApplicationContext` is the same data used in the `App` module to initialize the `WorkerProperties`, but it is now decoupled into an isolated module that helps to understand and maintain the application. In general, the `main` module of an application should be concerned with initializing application wide state. Also, as we mentioned earlier, the `App` class only needed to manage the `WorkerProperties` state so that it could coordinate this through from and to it's children.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "data used in the `App` module" -> "data that was used in the previous version of the `App` module"

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "application wide" -> "application-wide"

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "Also, as we mentioned earlier," -> "Also, as previously mentioned,"

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: "coordinate this through from and to it's children" too many pronouns, let's be more explicit... also would be its children

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+Now that we have the `ApplicationContext`, let's use it within our widgets. By creating an `Injector` and adding it to the `registry` created for the `projector`. `Injectors` are created using the `Injector` factory function from `@dojo/widget-core`.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "By creating an `Injector` and adding it to the `registry` created for the `projector`." Incomplete sentence, either missing a thought or ?

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+Now that the `Injector` is defined and registered and the `registry` has been set in the `projector`, it is time to create the components that will use it. In the next section, we will create a non-visual widget called a `Container` that will allow state to be injected into the `WorkerForm` and `WorkerContainer` widgets.
+
+{% section %}
+
+## Creating state containers
+
+{% task 'Create `Containers` that will allow state to be injected into widgets' %}
+
+On their own, `Injectors` are not able to help us very much because widgets expect state to be passed to them via properties. Therefore an `injector` must be connected to interested widgets in order for their state to be mapped to `properties` that widgets can consume, this is done using a `Container`. `Containers` are designed to coordinate the injection - they connect injectors to widgets and return `properties` from the `injectors` state which gets passed to the connected widgets.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "Therefore an `injector` must be connected to interested widgets in order for their state to be mapped to `properties` that widgets can consume, this is done using a `Container`. `Containers` are designed to coordinate the injection - they connect injectors to widgets and return `properties` from the `injectors` state which gets passed to the connected widgets." -> "Therefore an `injector` must be connected to interested widgets in order for their state to be mapped to `properties` that widgets can consume by using a `Container`. `Containers` are designed to coordinate the injection - they connect injectors to widgets and return `properties` from the `injector`'s state which is passed to the connected widgets."

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+Normally, a separate `Container` is created for each widget that needs to have `properties` injected. In the demo application, we have two widgets that rely on application state - `WorkerContainer` and `WorkerForm`. Let's start with the `WorkerContainer`. `Containers` are given the same name as the widget that they contain with the suffix `Container` appended. To keep things organized, they are also stored in a different sub-package - `containers`.
+
+{% instruction 'Add the following imports to the `WorkerContainerContainer` in the `containers` sub-package' %}
+
+{% include_codefile 'demo/finished/biz-e-corp/src/containers/WorkerContainerContainer.ts' lines:1-4 %}
+
+The first `import` gives the module access to the `Container` factory function which will be used to construct the container. The other `import` allow the module to use the `ApplicationContext` to extract state, use the `WorkerContainerProperties` to receive properties from its parent, and the `WorkerContainer` class that will be wrapped by the container.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "and the `WorkerContainer` class that will be wrapped by the container." -> "and wrap the `WorkerContainer` class with the `container`."

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+{% endaside %}
+
+The `getProperties` function receives two parameters. The first is the `payload` of the `injector` instance, for the demo application this is the data passed when creating the `ApplicationContext`. The second is the `properties` that have been passed to the container via the normal mechanism, `w(Container, properties)`, the properties will implement the properties interface defined by the wrapped widget (for example `WorkerContainerProperties`). The `getProperties` function must then return an object that holds the properties that will be passed to the widget itself. In this example, we are ignoring the properties provided by the parent and returning the `workerData` stored by the `ApplicationContext`. More advanced use cases where both sources are used to generate the properties are also possible.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "instance, for the" -> "instance. For the"

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: ", the properties will implement the properties interface" -> ". The properties will implement the properties interface"

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: ". The `getProperties` function" I would suggest this be the start of a new paragraph

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+
+The other container that we need is the `WorkerFormContainer`.
+
+{% instruction 'Add the following code to the `WorkerFormContainer` module in the `containers` sub-package.' %}
+
+{% include_codefile 'demo/finished/biz-e-corp/src/containers/WorkerFormContainer.ts' %}
+
+This module is almost identical to the `WorkerContainerContainer` except that additional properties are required by the `WorkerForm` to allow it to respond to user interactions with the form. The `ApplicationContext` contains two methods for managing these events - `onFormInput` and `onFormSave`. These methods need to be passed into the `WorkerForm` to handle the events, but they need to execute in the context of the `ApplicationContext`. To handle this, `bind` is called on each of the methods to explicitly set their execution contexts.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "except that additional properties are required by the" -> "except for additional properties that are required by the"

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+At this point, we have created the `ApplicationContext` to manage state, an `Injector` to inject state into the application's widgets, and `Containers` to manage how properties and children from the injector and parent widgets are combined. In the next section, we will integrate these components into our application.
+
+{% section %}
+
+## Using state containers
+
+{% task 'Integrate containers into an application.' %}
+
+As mentioned in the previous section, `Container` is a higher order component that extends `WidgetBase` and returns the wrapped widget and injected `properties` from the `render`. Given that, they can be used just like any other widget. In our demo application, we can take advantage of that fact by simply replacing the `WorkerForm` and `WorkerContainer` with their container equivalents.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "Given that, " -> "As such, "

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: "we can take advantage of that fact by " -> "we can take advantage of their extension of `WidgetBase` by "

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+{% instruction 'Replace the imports in the `App` module with the following.' %}
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' lines:1-6 %}
+
+There are two major changes to the `App` module's imports. First, the widgets (`WorkerForm` and `WorkerContainer`) have been replaced by their container equivalents (`WorkerFormContainer` and `WorkerContainerContainer`). Second, all of the objects related to state management have been removed, `WorkerFormData`, and `WorkerProperties`. These are no longer needed since the `App` class no longer needs to manage state.
+
+The new class declaration for `App` should be:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' line:8 %}
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: Something is wrong with this statement as a blank line is rendered here

File: site/source/tutorials/1010_containers_and_injecting_state/index.md [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+
+The final change to `App` is to update the `render` method to use the containers. Since the containers already know how to manage their state and respond to events, no properties need to be passed directly to the `Container` by the `App` widget.
+
+{% instruction 'Replace the `render` method with the following code.' %}
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' lines:9-15 %}
+
+With this last change, the `App` class is now only nine lines of code. All of the state management logic is still part of the application, but it has been refactored out of the `App` class to create a more efficient application architecture.
+
+Notice that the `WorkerForm` and `WorkerContainer` widgets were not changed at all. This is an important thing to keep in mind when designing widgets - a widget should never be tightly coupled to the source of its properties. By keeping the containers and widgets separate, we have helped to ensure that they each have a narrowly defined set of responsibilities, creating a cleaner separation of concerns, and making the widgets and containers.
```
Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "were not changed at all." I think this deserves to be "were not changed at all!" as it's impressive

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "that they each have " -> "the each widget or container has"

Comment (dylans) [inline site/source/tutorials/1010_containers_and_injecting_state/index.md]: "and making the widgets and containers." incomplete final thought?

File: site/source/tutorials/1010_containers_and_injecting_state/index.md (hunk 8 above)
Comment (agubler) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: Change to "lifting up state", it is a pattern that is familiar in the community (ala React)

File: site/source/tutorials/1010_containers_and_injecting_state/index.md (hunk 20 above)
Comment (agubler) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: @dylans ha, that come from your comment on the original PR - so do you have a final thought? 😄

Comment (dylans) [reply site/source/tutorials/1010_containers_and_injecting_state/index.md]: How about "creating a cleaner separation of concerns within our widgets and containers"?

**Comment by agubler** [comment]
@dylans is this alright to be immediately published to dojo.io? I think that will happen automatically on merge?

**Comment by dylans** [comment]
I think it's ready to go live, no reason to hold it back.
